File: C#/122. Best Time to Buy and Sell Stock II.cs
class Solution
{
    public:
    int maxProfit(vector<int>& prices)
    {
        int n = prices.size();
        if (n <= 1) return 0;
        vector<int> dp(n,0);
        int minPrice = prices[0], res = 0;
        for (int i = 1; i < n; ++i)
        {
            minPrice = min(minPrice, prices[i]);
            if (prices[i] - minPrice >= dp[i - 1]) dp[i] = prices[i] - minPrice;
            else
            {
                minPrice = prices[i];
                res += dp[i - 1];
            }
        }

        return res + dp[n - 1];
    }
};

File: python/122. Best Time to Buy and Sell Stock II.py
# -*- coding: utf-8 -*-

"""
@Author: xiezizhe
@Date: 27/2/2020 上午12:29
"""
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if prices is None or len(prices) <= 1:
            return 0

        days = len(prices)
        buy, nonbuy = [0] * days, [0] * days
        buy[0] = -prices[0]

        for i in range(1, days):
            buy[i] = max(nonbuy[i - 1], buy[i - 1] + prices[i]) - prices[i]
            nonbuy[i] = max(nonbuy[i - 1], buy[i - 1] + prices[i])

        return nonbuy[-1]


if __name__ == "__main__":
    s = Solution()
    print(s.maxProfit([7, 1, 5, 3, 6, 4, 6, 7, 5, 4, 1]))  # 10
    print(s.maxProfit([4, 6, 7]))  # 3
